Add user methods

damionvega opened the pull request:
Closes #11 

I added a quick `refresh` function to illustrate how the `user.update()` method would work, the test can remove this piece after we decide on the final format.

## Overview
We need a user object to work with after `Userfront` has been initialized and the tokens are set.

When the tokens are set via `setCookiesAndTokens`, a new function `setUser` is called which creates a user object:

```js
{
  // User information
  email: 'johndoe@example.com',
  username: 'johndoe',
  name: 'John Doe',
  image: 'https://example.com/profile-image.png',
  confirmedAt: '2021-05-19T19:12:45.150Z',
  createdAt: '2021-05-19T19:12:45.150Z',
  updatedAt: '2021-05-19T19:12:45.150Z',
  age: 22,
  color: 'blue',
  birthdate: '2021-05-19T19:12:45.150Z',
  data: { country: 'Monoco' },
  metadata: { lastLogin: '2021-05-19T19:12:45.150Z' },
  userId: 3,
  authorization: { abcdefgh: { roles: [Array] } },

  // `update` method
  update: [AsyncFunction: update]
}
```

Which gives us the following API:

```js
const { user } = Userfront.store

console.log(user.name) // John Doe

// Update name
await user.update({ name: 'Jane Doe' })
```

The `update` method belongs to the `apiResource` object which is used to separate networking from the user object. One of the options of `apiResource` is an `afterUpdate` a callback to call after `user.update` is called, which we can use to call `Userfront.refresh`.

```js
apiResource({
  store,
  path: "/users",
  id: userId,
  afterUpdate: refresh,
})
```

The `update` method allows for future simplified functions against our basic user columns:

```js
const { user } = Userfront.store

await user.updateImage(url)
await user.updateName('Jane Doe')
```

## Other thoughts
- Having a method on the Userfront object itself didn't really make sense because there isn't a reason to update anything about a user without `Userfront.user`. With that said, I went against something simpler like:

    ```js
    await Userfront.updateUserName('Jane Doe')
    await Userfront.updateUserImage(url)
    
    // or with general update method
    await Userfront.updateUser({
     name: 'Jane Doe',
     image: url
    })
    ```
- I decided to go with factory functions over classes simply for flexibility and the fact there aren't going to be many instances of the user object or "user api resource" object.
- The `apiResource` provides the ability to work with other resources in the future:

    ```js
    // In `setUser`:
    store.user.roles = apiResource({
      path: `/users/${userId}/roles`,
      afterCreate: refresh,
    })

    // Usage:
    const { user } = Userfront.store
    console.log(user.authorization['abcdefgh']) // { roles: ['member'] }

    await user.roles.create('subscriber')
    console.log(user.authorization['abcdefgh']) // { roles: ['member', 'subscriber'] }

    await user.roles.delete('member')
    console.log(user.authorization['abcdefgh']) // { roles: ['subscriber'] }

    await user.roles.update(['admin', 'viewer'])
    console.log(user.authorization['abcdefgh']) // { roles: ['admin', 'viewer'] }
    ```

damionvega commented on src/index.js:
```diff
@@ -423,6 +430,47 @@ function setCookiesAndTokens(tokens) {
+  setUser();
+}
+
+/**
+ * Define user attributes & methods based on access & ID token.
+ */
+function setUser() {
+  if (!store.idToken) {
+    throw new Error("ID token has not been set.");
+  }
+  if (!store.accessToken) {
+    throw new Error("Access token has not been set.");
+  }
+
+  const decodedIdToken = jwt.decode(store.idToken);
+  const decodedAccessToken = jwt.decode(store.accessToken);
+
+  store.user = {
+    ...user({ idToken: decodedIdToken, accessToken: decodedAccessToken }),
+    ...apiResource({
+      store,
+      path: "/users",
+      id: decodedAccessToken.userId,
+      afterUpdate: refresh,
+    }),
+  };
+}
```
Here is where `Userfront.store.user` is created.

- `...user` is the user information from the tokens
- `...apiResource` is what provides the `user.update` method
  The `afterUpdate` callback option was created in order to keep the `refresh` in one place and leave room to do other work if necessary.

damionvega commented on src/index.js:
```diff
@@ -423,6 +430,47 @@ function setCookiesAndTokens(tokens) {
+  };
+}
+
+async function refresh() {
+  const { data } = await axios.get({
+    url: `${apiUrl}tenants/${store.tenantId}/refresh`,
+    headers: {
+      authorization: `Bearer ${store.accessToken}`,
+    },
+  });
+
+  if (data.tokens) {
+    setCookiesAndTokens(data.tokens);
+  } else {
+    throw new Error("Problem refreshing tokens.");
+  }
```
This can be modified/thrown away. Is only here to satisfy the current tests (`setCookiesAndTokens` needs to be called in order to update the tokens).

damionvega commented on src/api-resource.js:
```diff
@@ -0,0 +1,60 @@
+ * @property {Function} options.afterUpdate Function to call after update()
+ * @property {Function} options.afterDelete Function to call after delete()
+ */
+export default (options = {}) => {
+  const { store, path, id } = options;
+  const { tenantId, accessToken } = store;
+  if (!tenantId) {
+    throw new Error("API resource error: Missing tenant ID");
+  }
+  if (!accessToken) {
+    throw new Error("API resource error: Missing access token");
+  }
+  if (!path) {
+    throw new Error("API resource error: Missing path");
+  }
+
+  let { basePath } = options;
+  if (!basePath) {
+    basePath = `${apiUrl}tenants/${tenantId}`;
+  }
```
Currently `tenantId` and `accessToken` are required. If we decide to go forward with the `apiResource` design, I'd probably make another factory function e.g. `apiBase` (not sure of the name) that `apiResource` could adopt functionality from here and remove the need to pass those as options.

tyrw commented:
Can you clarify:
- As a project admin, how do I set a value?  E.g. `image` and `firstName`
- As a project admin, how do I read a value?  E.g. `image` and `firstName`

damionvega commented:
```js
const { user } = Userfront.store

// Read image value
user.image

// Read firstName value (is `undefined` if not present on `data` object in ID token)
user.firstName

// Update image and firstName
await user.update({
  image: 'https://example.com/profile-image.png',
  data: {
    firstName: 'Jane'
  }
})
```

- `user.updateImage` function can be added
- Updating attributes that require adding `data: { ... }` can be removed in order to allow for a flat object:
  ```js
  await user.update({
    image: 'https://example.com/updated.jpg',
    firstName: 'Jane',
  })
  ```
- If `user.firstName` is in popular demand and isn't set from the `data` object, perhaps we create a virtual property that returns the characters before before the first space in `user.name`.

tyrw commented:
From the chat:

Can you make this so the user doesn't care about implementation details:

```
await user.update({
  image: 'https://example.com/profile-image.png',
  data: {
    firstName: 'Jane'
  }
})
```

Should be

```
await user.update({
  image: 'https://example.com/profile-image.png',
  firstName: 'Jane'
})
```

As is, `user.firstName` doesn't make sense if they had to set it with `user.data.firstName`

It might be best to flatten data into the id token overall

E.g. the project admin would never need to know about `data`

tyrw commented on src/user.js:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Create a user object based on tokens
+ * @param {Object} tokens
+ * @property {Object} idToken Decoded ID token
+ * @property {Object} accessToken Decoded access token
+ * @returns {Object}
+ */
```
The access token should not be needed for constructing the `user` object.  Looks like you are only using it for `authorization`, which we shouldn't commingle.

tyrw reviewed (changes requested):
The structure of the code seems to have a system to it, but it's hard to tell what you are going for.  I think this is partly due to lack of documentation about architecture or high-level API, partly due to confusing tests, and partly due to lack of smaller comments in tests.

- Please document somewhere what the syntax is
- The tests all reference `Userfront.store.user`, but you said it is `Userfront.user`. Can you write the tests to use that instead?
- This implementation doesn't seem to use the JWKS to verify the ID token. Am I missing it?
- It would be really nice if the tests were more descriptive of how things work, following a format more similar to our backend test suite: setup, action, assertion. E.g. a test might look look like:

```js
describe('user.update()', () => {
  before(() => {
    // Setup and mocks
  })
  it('should update a user', async () => {
    // Existing state
    expect(Userfront.user.image).toEqual('https://example.com/start.png')
    expect(Userfront.user.firstName).toBe(undefined)
    
    // Action
    const newImage = 'https://example.com/update.png'
    const newFirstName = 'Johnny'
    await Userfront.user.update({ 
      image: newImage,
      firstName: newFirstName
    })

    // Should have updated the user object
    expect(Userfront.user.image).toEqual(newImage)
    expect(Userfront.user.firstName).toEqual(newFirstName)

    // Should have updated the id token
    const idPayload = jwt.verify(Userfront.idToken(), publicKey)
    expect(idPayload.image).to.equal(newImage)
    expect(idPayload.firstName).to.equal(newFirstName)

    // Should have made the API request correctly
    ...
  })
})
```

tyrw commented:
Also `api-resource.js` and "The `apiResource` design" both come across as very vague.  I read and re-read, and I still don't know what it's supposed to mean or reference.

damionvega commented:
I've added `verifyToken(token)` and `setUser()`. Since we currently have a single "model" we're working with, I decided to keep things simple and refactored the `apiResource`'s `update` function to the `user` factory and removed `api-resource.js`.

### verifyToken - public (`Userfront.verifyToken()`)
Primarily used to verify ID token before setting `Userfront.user`. Made available for admin to use if desired.
1. Decodes provided token
2. Fetches JWKS from `/jwks/{mode}`,
3. Gets public key from matching JWK
4. Verifies JWT with public key

### setUser - private
Called after `store.idToken` is set. Verifies the ID token and sets `store.user` for admin's use
of `Userfront.user`.
1. Verifies ID token via `verifyToken`
2. Assigns user object to `store.user` based on information in ID token and includes an `update` method to update user information

### Current API
```js
// Given example data:
{
  email: 'johndoe@example.com',
  username: 'johndoe',
  name: 'John Doe',
  image: 'https://example.com/profile-image.png',
  data: {
    country: 'Monoco',
  }
  // ...other properties
}

console.log(Userfront.user.name) // John Doe
console.log(Userfront.user.data.country) // Monoco

// Update user via Userfront API
await Userfront.user.update({
  username: "john-doe-updated",
  data: {
    country: "Spain",
  },
});

console.log(Userfront.user.username) // john-doe-updated
console.log(Userfront.user.data.country) // Spain
```

tyrw commented on src/index.js:
```diff
@@ -425,6 +475,44 @@ function setCookiesAndTokens(tokens) {
+    throw error;
+  }
+  console.log("Setting user", store);
+  store.user = user({
+    store,
+    afterUpdate: refresh,
+  });
```
Assignment with `=` destroys any existing references to `user` that have been exported as

```js
const { user } = Userfront
```

tyrw commented on test/user.spec.js:
```diff
@@ -0,0 +1,188 @@
+
+import utils from "./utils.js";
+import { apiUrl } from "../src/constants.js";
+import Userfront from "../src/index.js";
+import userConstructor from "../src/user.js";
+
+const { user } = Userfront;
```
Please import into all the tests like this, so that we are testing the way it will be used by the admins.

tyrw reviewed (changes requested):
I updated the tests so that the `user` object is imported as it will be by others, rather than importing the `/src/user.js` file and testing on that.  This distinction is important, because it's how admins will use the package, so it needs to not be broken.

I made a note below about why it is broken, and it should be a straightforward fix, but I wanted to put it to you so that it clicks for future work.

tyrw commented:
@damionvega the live version of this is available at https://mod.userfront.com/vtest, which means you can test it by replacing `/v3` in a snippet with `/vtest`, like so: 

```html
<script id="Userfront-script">
  (function(m,o,d,u,l,a,r,i,z,e) {
    u[m]={rq:[],ready:function(j){u[m].rq.push(j);},m:m,o:o,d:d,r:r};function j(s){return encodeURIComponent(btoa(s));}z=l.getElementById(m+"-"+a);r=u.location;
    e=[d+"/page/"+o+"/"+j(r.pathname)+"/"+j(r.host)+"?t="+Date.now(),d];e.map(function(w){i=l.createElement(a);i.defer=1;i.src=w;z.parentNode.insertBefore(i,z);});u.amvartem=m;
  })("Userfront", "p9ny8bdj", "https://mod.userfront.com/vtest",window,document,"script");
</script>
```

2 things to note:

- There are errors.  See console of https://codesandbox.io/s/recursing-grass-x4s46?file=/index.html for an example
- The bundle file is huge.  It went from 18kB without `jwks-rsa` to 287kB with it.  This is probably a deal breaker if there's no workaround.
